ASoC: Intel: select SND_SOC_SDCA by SND_SOC_ACPI_INTEL_SDCA_QUIRKS

bardliao opened the pull request:
SND_SOC_SDCA can't be optional when SND_SOC_ACPI_INTEL_SDCA_QUIRKS is selected. IS_REACHABLE can prevent the link error. However it is not suitable for this case. When CONFIG_SND_SOC_ACPI_INTEL_SDCA_QUIRKS is Y and CONFIG_SND_SOC_SDCA is M, the SDCA helpers will be empty and return false. But we need the helpers to do their jobs whenSND_SOC_SDCA is M.

Suggested-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

bardliao committed:
ASoC: Intel: Kconfig: select SND_SOC_SDCA by SND_SOC_ACPI_INTEL_SDCA_QUIRKS

SND_SOC_SDCA can't be optional when SND_SOC_ACPI_INTEL_SDCA_QUIRKS is
selected. IS_REACHABLE can prevent the link error. However it is not
suitable for this case. When CONFIG_SND_SOC_ACPI_INTEL_SDCA_QUIRKS is
Y and CONFIG_SND_SOC_SDCA is M, the SDCA helpers will be empty and
return false. But we need the helpers to do their jobs whenSND_SOC_SDCA
is M.
IOW, the SDCA library is not optional for Intel platforms where the
SDCA_QUIRK is selected.
Also, make SND_SOC_SDCA invisible. SND_SOC_SDCA should be selected if
a device supports SDCA. User should not unselect it.

Reported-by: kernel test robot <lkp@intel.com>
Closes: https://lore.kernel.org/oe-kbuild-all/202411021722.DiX1Y5sf-lkp@intel.com/
Suggested-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on sound/soc/intel/Kconfig:
```diff
@@ -77,7 +77,7 @@ config SND_SOC_ACPI_INTEL_MATCH
 config SND_SOC_ACPI_INTEL_SDCA_QUIRKS
 	tristate
-	imply SND_SOC_SDCA
+	select SND_SOC_SDCA
```
Can you also add:

Reported-by: kernel test robot [<lkp@intel.com>](mailto:lkp@intel.com)
Closes: https://lore.kernel.org/oe-kbuild-all/202411021722.DiX1Y5sf-lkp@intel.com/

ujfalusi replied:
iow, the SDCA library is not optional for INtel platforms where the SDCA_QUIRK is selected.

bardliao replied:
> Can you also add:
> 
> Reported-by: kernel test robot [lkp@intel.com](mailto:lkp@intel.com) Closes: https://lore.kernel.org/oe-kbuild-all/202411021722.DiX1Y5sf-lkp@intel.com/

Of course, thanks for reminding me.

lgirdwood replied:
@ujfalusi @bardliao This makes me think SDCA library should be default=y for Intel in the near term. i.e. we should plan this soon.

charleskeepax replied:
I believe if you change this to a select you also need to pull forward the depends, since select will ignore the depends of its target. So you probably need to add a depends on ACPI into SND_SOC_ACPI_INTEL_SDCA_QUIRKS.

ujfalusi replied:
and to add, the advice is that the select should not be used for user visible options...

SND_SOC_ACPI_INTEL_MATCH is only selected if ACPI is enabled, but randconfig might be able to generate a config which can fail.

ujfalusi replied:
might be this?
```diff
diff --git a/sound/soc/intel/Kconfig b/sound/soc/intel/Kconfig
index 14461dee3e52..46b45f390ae9 100644
--- a/sound/soc/intel/Kconfig
+++ b/sound/soc/intel/Kconfig
@@ -70,14 +70,15 @@ if SND_SOC_INTEL_SST_TOPLEVEL || SND_SOC_SOF_INTEL_TOPLEVEL
 
 config SND_SOC_ACPI_INTEL_MATCH
 	tristate
-	select SND_SOC_ACPI if ACPI
+	depends on ACPI
+	select SND_SOC_ACPI
 	select SND_SOC_ACPI_INTEL_SDCA_QUIRKS
 	# this option controls the compilation of ACPI matching tables and
 	# helpers and is not meant to be selected by the user.
 
 config SND_SOC_ACPI_INTEL_SDCA_QUIRKS
 	tristate
-	imply SND_SOC_SDCA
+	select SND_SOC_SDCA
 
 endif ## SND_SOC_INTEL_SST_TOPLEVEL || SND_SOC_SOF_INTEL_TOPLEVEL
 ```
Randconfig can still set SND_SOC_ACPI_INTEL_SDCA_QUIRKS w/o SND_SOC_ACPI_INTEL_MATCH?

bardliao replied:
Sounds good to me. I don't know how can SND_SOC_ACPI_INTEL_MATCH work without ACPI. Also, I think we can make SND_SOC_SDCA invisible in this PR.

bardliao committed:
ASoC: Intel: Kconfig: make SND_SOC_ACPI_INTEL_MATCH depend on ACPI

SND_SOC_ACPI_INTEL_MATCH relies on ACPI functions. It will not work
if ACPI is not selected.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ranj063 commented:
@bardliao is the compress_offload build warning new?

bardliao commented:
> @bardliao is the compress_offload build warning new?

No,
```
sound/core/compress_offload.c:409:35: error: variable ‘runtime’ set but not used [-Werror=unused-but-set-variable]
  409 |         struct snd_compr_runtime *runtime;
      |                                   ^~~~~~~
 ```
 is not new

ujfalusi commented:
> > @bardliao is the compress_offload build warning new?
> 
> No,
> 
> ```
> sound/core/compress_offload.c:409:35: error: variable ‘runtime’ set but not used [-Werror=unused-but-set-variable]
>   409 |         struct snd_compr_runtime *runtime;
>       |                                   ^~~~~~~
> ```
> 
> is not new

@bardliao, @ranj063, https://lore.kernel.org/linux-sound/20241106070646.2599-1-peter.ujfalusi@linux.intel.com/
